### ISSUE TITLE
Faster Parsing

### ISSUE THREAD

**Issue by chriseidhof** [davedufresne/SwiftParsec#6]
Hey @davedufresne !

We managed to build a working parser for something like ledger-cli. The 5000-line sample file that we tried to parse (naively) took a massive amount of time to parse (the first iteration even took 13 minutes). We got it down to a few minutes.

The `popFirst()` on strings is really expensive, as it reallocates a new copy of the string. We managed to speed it up a little bit by using a `CharacterView`, but still, this mutates the underlying `_StringCore` type.

As `Stream` only needs the `ArrayLiteralConvertible` and the `popFirst()`, we implemented a new struct that doesn't copy at all. It only changes the "pointer" to the first element in the array. We also tried storing `String.CharacterView` in this struct, but that was a lot slower.

```swift
struct ImmutableCharacters: Stream {
    var characters: [Character]
    var start: Int = 0
    
    init(string: String) {
        characters = Array(string.characters)
    }
    
    init(arrayLiteral elements: Character...) {
        characters = elements
    }
    
    mutating func popFirst() -> Character? {
        guard start < characters.count else { return nil }
        let oldStart = start
        start += 1
        return characters[oldStart]
    }
}
```

This helped a lot for speed. Our file now parses in 2 seconds. This is still quite slow (we'd like to be able to parse it in milliseconds).

When running this in Instruments, we noticed at least half a second was spent retaining/releasing things. Almost certainly, the `characters` array. If could work on something like an `UnsafeBufferPointer`, there would be no need for retain or release. However, this is hard to make work with both the `Parsec.string` parser, and also the `ArrayLiteralConvertible` that `Stream` inherits from.

Have you thought about these issues, and do you have any more ideas for how to make this fast? We're happy to help out. Maybe we could integrate the `ImmutableCharacters` struct already? It's a drop-in replacement.

**Comment by davedufresne** [comment]
Hey Chris,

Many thanks for you feedback, it's really appreciated!

The `ImmutableCharacters` struct is basically an iterator, which is a good idea. We could generalize your concept by using an iterator provided by a type implementing the `Collection` protocol. I quickly looked at the code, so I'm not sure about all the implications, but here is what could be done:

- Get away with the `Stream` protocol and replace it with the `Collection` Protocol. So the new type constraint for the `Parsec` `StreamType` associated type would be like that: `associatedtype StreamType: Collection`.
- Instead of passing around the whole collection, only use the iterator (`input.makeIterator()`).
- `input.popFirst()` would be replaced by `input.next()`.

The advantage of this solution is that the code remains generic.

Tell me what you think. I'll be able to look into this issue a bit further this weekend.

**Comment by chriseidhof** [comment]
That sounds like a good plan. Does it need to be a collection, or could it be just an `Iterator`? We could just "destructively" consume the iterator, right? Or do we need value semantics and repeated access?

**Comment by davedufresne** [comment]
The input cannot be destructively consumed because of the `attempt` parser that has to backtrack on failure. That's why I suggested the `Collection` protocol, it's multi-pass.

**Comment by chriseidhof** [comment]
Of course =). BTW, if you want to chat about this stuff on Skype or via im
sometime, let me know. It should be easy to find me ;).

On Thu, Jun 30, 2016 at 1:59 PM, David Dufresne <notifications@github.com>
wrote:

> The input cannot be destructively consumed because of the attempt parser
> that has to backtrack on failure. That's why I suggested the Collection
> protocol, it's multi-pass.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/davedufresne/SwiftParsec/issues/6#issuecomment-229637955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAVBhDTUq4Dd6cGgeWFv9RamPYI4OsZks5qQ6-FgaJpZM4JAN9f>
> .
>



-- 
Chris Eidhof

**Comment by davedufresne** [comment]
I fixed the problem. Now the collection being parsed is traversed using an iterator. So the collection is not mutated during the parsing. It didn't change the API, the changes are internal only. They are in the 2.0-dev branch. Could you tell me how much it improved the speed. After we could check if there is more improvements to make.

**Comment by chriseidhof** [comment]
I haven't checked out the branch yet, will do that in a bit. Some things I've learned while trying to optimize our parser over the weekend (I started with a test that took 2.5 seconds to run):

When using the `ImmutableCharacters` struct, I managed to get a big speedup making the `tokens` method more generic. Instead of it taking a `StreamType`, it can now take any `Stream`-conforming type, as long as the elements match. That saves a lot of conversions. The `string` function I use converts the `String` into a character array first before using the built-in `.string` method. Now the test took 1.8 seconds.

I extended the `Stream` protocol to include a `first` computed property, which peeks (instead of mutating the input). I then added a combinator `onlyIf` which only succeeds if the `first` element matches a condition. This saves us from building/destroying results, and gives us quite a speedup (might ba a quite specific use-case). It allowed us to remove almost all `attempt`s. (1.5 seconds)

```swift
public extension GenericParser where StreamType.Element: Equatable {    
    public func onlyIf(peek check: (StreamType.Element) -> Bool) -> GenericParser<StreamType, UserState, Result> {
```

Finally, I changed the implementation of `satisfy` to `return anyToken.onlyIf(peek: predicate)` and got another decent speedup (tests take 1.2 seconds).

I'll have a look at your changes later on!
